File: apps/macro-sandbox/functions/javascript/handler.js
const { execFile } = require("child_process");
const fs = require("fs");
const path = require("path");
const os = require("os");
const zlib = require("zlib");

const { guardBatch, partitionItems, mergeResults } = require(
  path.join(__dirname, "guards", "guard.js"),
);

// Shadow classifies and logs only; enforce rejects non-canonical items. Shadow
// is the default so this release does not change execution (Ticket 02).
function guardMode() {
  return process.env.MACRO_GUARD_MODE === "enforce" ? "enforce" : "shadow";
}

// AWS Lambda sync responses are capped at 6 MB. Compress every response so
// macro outputs of ~25-50 MB raw can still fit. Callers detect the
// {encoding, payload} wrapper and decompress.
function compressResponse(envelope) {
  const json = JSON.stringify(envelope);
  return {
    encoding: "gzip+base64",
    payload: zlib.gzipSync(json).toString("base64"),
  };
}

// Limits
const MAX_SCRIPT_SIZE = 1 * 1024 * 1024; // 1MB
const MAX_ITEM_COUNT = 1000;
const MAX_TIMEOUT = 60;
const DEFAULT_TIMEOUT = 10;

const WRAPPER_PATH = "/var/task/wrappers/wrapper.js";

// Warm-start cleanup: remove leftover temp dirs from crashed invocations.
function cleanupStaleTmp() {
  try {
    const tmpBase = os.tmpdir();
    for (const entry of fs.readdirSync(tmpBase)) {
      if (entry.startsWith("macro_")) {
        fs.rmSync(path.join(tmpBase, entry), { recursive: true, force: true });
      }
    }
  } catch {
    // best-effort
  }
}

exports.handler = async (event) => {
  const result = await _runMacroBatch(event);
  return compressResponse(result);
};

async function _runMacroBatch(event) {
  let tmpdir;

  try {
    cleanupStaleTmp();
    // Validate script
    if (!event.script) {
      return { status: "error", results: [], errors: ["Missing 'script' field"] };
    }

    // Node's Buffer.from silently ignores invalid base64 chars; round-trip check catches corruption.
    const scriptBytes = Buffer.from(event.script, "base64");
    if (scriptBytes.toString("base64") !== event.script) {
      return { status: "error", results: [], errors: ["Invalid base64 in 'script'"] };
    }
    if (scriptBytes.length > MAX_SCRIPT_SIZE) {
      return { status: "error", results: [], errors: ["Script exceeds 1MB limit"] };
    }
    const scriptContent = scriptBytes.toString("utf8");

    // Validate items
    const items = event.items || [];
    if (!Array.isArray(items)) {
      return { status: "error", results: [], errors: ["'items' must be an array"] };
    }
    if (items.length > MAX_ITEM_COUNT) {
      return {
        status: "error",
        results: [],
        errors: [`Exceeds ${MAX_ITEM_COUNT} item limit`],
      };
    }

    const timeout = Math.max(
      DEFAULT_TIMEOUT,
      Math.min(parseInt(event.timeout) || DEFAULT_TIMEOUT, MAX_TIMEOUT),
    );

    // Contract guard. Shadow classifies + logs; execution is unchanged.
    const mode = guardMode();
    const guard = guardBatch({ input_contract: event.input_contract, items }, mode);
    console.error("[guard] " + JSON.stringify(guard.telemetry));

    let execItems = items;
    let invalidResults = [];
    if (mode === "enforce") {
      if (!guard.markerValid) {
        return { status: "error", results: [], errors: [guard.markerError] };
      }
      const partitioned = partitionItems(items, guard.decisions);
      execItems = partitioned.validItems;
      invalidResults = partitioned.invalidResults;
    }

    // Write temp files
    tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), "macro_"));
    const scriptPath = path.join(tmpdir, "script");
    const inputPath = path.join(tmpdir, "input.json");

    fs.writeFileSync(scriptPath, scriptContent, { mode: 0o600 });
    fs.writeFileSync(inputPath, JSON.stringify(execItems), { mode: 0o600 });

    // Run wrapper in a subprocess with a stripped environment.
    const result = await new Promise((resolve) => {
      execFile(
        "node",
        [WRAPPER_PATH, scriptPath, inputPath],
        {
          timeout: (timeout + 5) * 1000,
          env: {
            PATH: "/var/lang/bin:/usr/local/bin:/usr/bin:/bin",
            HOME: "/tmp",
            NODE_PATH: "/var/task",
          },
          maxBuffer: 10 * 1024 * 1024, // 10MB output buffer
        },
        (error, stdout, stderr) => {
          // maxBuffer overflow can fire without setting error.killed/error.signal,
          // so check the code first.
          if (error?.code === "ERR_CHILD_PROCESS_STDIO_MAXBUFFER") {
            resolve({
              status: "error",
              results: [],
              errors: ["Wrapper output exceeds 10MB limit"],
            });
            return;
          }
          if (error && error.killed) {
            resolve({
              status: "error",
              results: [],
              errors: ["Execution timed out"],
            });
            return;
          }

          const output = (stdout || "").trim();
          if (output) {
            try {
              resolve(JSON.parse(output));
            } catch (parseErr) {
              const code = error?.code ?? "unknown";
              const signal = error?.signal ?? "none";
              // Stderr goes to CloudWatch via console.error, not into the
              // response. The response propagates into the macro_error column
              // downstream and shouldn't carry server internals.
              const stderrTrimmed = (stderr || "").trim();
              if (stderrTrimmed) {
                console.error("[handler] wrapper stderr:", stderrTrimmed.slice(0, 4000));
              }
              resolve({
                status: "error",
                results: [],
                errors: [
                  `Wrapper returned invalid JSON (parse: ${parseErr.message}; exit: ${code}; signal: ${signal}; stdoutLen: ${output.length}; stderrLen: ${stderrTrimmed.length})`,
                ],
              });
            }
          } else {
            resolve({
              status: "error",
              results: [],
              errors: ["Wrapper returned no output"],
            });
          }
        },
      );
    });

    // Reassemble pre-failed items with wrapper results in request order.
    if (mode === "enforce" && result.status === "success") {
      if (!Array.isArray(result.results) || result.results.length !== execItems.length) {
        return {
          status: "error",
          results: [],
          errors: ["Wrapper result count mismatch"],
        };
      }
      return {
        status: "success",
        results: mergeResults(guard.decisions, result.results, invalidResults),
      };
    }

    return result;
  } catch (e) {
    return {
      status: "error",
      results: [],
      errors: [`Handler error: ${e.constructor.name}`],
    };
  } finally {
    if (tmpdir) {
      fs.rmSync(tmpdir, { recursive: true, force: true });
    }
  }
}

File: apps/macro-sandbox/functions/r/handler.R
#!/usr/bin/env Rscript
# Lambda handler for R macro execution.
# Called by the bootstrap: Rscript handler.R <event_file>

suppressPackageStartupMessages({
  library(jsonlite)
})

# Limits
MAX_SCRIPT_SIZE <- 1048576  # 1MB
MAX_ITEM_COUNT  <- 1000
MAX_TIMEOUT     <- 60
DEFAULT_TIMEOUT <- 10

WRAPPER_PATH <- "/var/task/wrappers/wrapper.R"

# Read event file path from command args
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  cat(toJSON(list(status = "error", results = list(), errors = list("No event file provided")), auto_unbox = TRUE))
  quit(status = 0)
}

event_file <- args[1]

# Parse event
event <- tryCatch({
  fromJSON(event_file, simplifyVector = FALSE)
}, error = function(e) {
  cat(toJSON(list(status = "error", results = list(), errors = list(paste0("Failed to parse event: ", e$message))), auto_unbox = TRUE))
  quit(status = 0)
})

# Validate script
if (is.null(event$script)) {
  cat(toJSON(list(status = "error", results = list(), errors = list("Missing 'script' field")), auto_unbox = TRUE))
  quit(status = 0)
}

# Decode base64 script
script_bytes <- tryCatch({
  base64_dec(event$script)
}, error = function(e) {
  cat(toJSON(list(status = "error", results = list(), errors = list("Invalid base64 in 'script'")), auto_unbox = TRUE))
  quit(status = 0)
})

if (length(script_bytes) > MAX_SCRIPT_SIZE) {
  cat(toJSON(list(status = "error", results = list(), errors = list("Script exceeds 1MB limit")), auto_unbox = TRUE))
  quit(status = 0)
}

script_content <- rawToChar(script_bytes)

# Validate items
items <- event$items
if (is.null(items)) items <- list()
if (!is.null(names(items))) {
  cat(toJSON(list(status = "error", results = list(), errors = list("'items' must be an array, not an object")), auto_unbox = TRUE))
  quit(status = 0)
}
if (length(items) > MAX_ITEM_COUNT) {
  cat(toJSON(list(status = "error", results = list(), errors = list(paste0("Exceeds ", MAX_ITEM_COUNT, " item limit"))), auto_unbox = TRUE))
  quit(status = 0)
}

timeout <- max(DEFAULT_TIMEOUT, min(as.numeric(ifelse(is.null(event$timeout), DEFAULT_TIMEOUT, event$timeout)), MAX_TIMEOUT))

# Warm-start cleanup: remove leftover temp dirs from crashed invocations.
for (d in Sys.glob("/tmp/macro_*")) {
  unlink(d, recursive = TRUE)
}

# Write temp files
tmpdir <- tempfile(pattern = "macro_", tmpdir = "/tmp")
dir.create(tmpdir)
script_path <- file.path(tmpdir, "script")
input_path <- file.path(tmpdir, "input.json")
output_path <- file.path(tmpdir, "output.json")

writeLines(script_content, script_path)
writeLines(toJSON(items, auto_unbox = TRUE), input_path)

# Run wrapper in a subprocess with stripped environment (env -i).
# Output file (3rd arg) prevents user cat()/print() from corrupting JSON.
result <- tryCatch({
  exit_code <- system2(
    "env",
    args = c(
      "-i",
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
      "HOME=/tmp",
      "timeout", as.character(timeout + 5),
      "Rscript", WRAPPER_PATH, script_path, input_path, output_path
    ),
    stdout = FALSE,
    stderr = FALSE
  )

  if (exit_code == 124) {
    list(status = "error", results = list(), errors = list("Execution timed out"))
  } else if (file.exists(output_path)) {
    output_size <- file.info(output_path)$size
    if (output_size > 10 * 1024 * 1024) {
      list(status = "error", results = list(), errors = list("Wrapper output exceeds 10MB limit"))
    } else {
      output_str <- paste(readLines(output_path, warn = FALSE), collapse = "\n")
      if (nchar(output_str) > 0) {
        fromJSON(output_str, simplifyVector = FALSE)
      } else {
        list(status = "error", results = list(), errors = list("Wrapper returned no output"))
      }
    }
  } else {
    list(status = "error", results = list(), errors = list("Wrapper produced no output file"))
  }
}, error = function(e) {
  list(status = "error", results = list(), errors = list(paste0("Execution failed: ", e$message)))
})

# Clean up
unlink(tmpdir, recursive = TRUE)

# Produce true gzip bytes (with 0x1f 0x8b magic) for the given text.
# memCompress(., "gzip") emits zlib-format output despite the name, which
# Node's gunzipSync rejects ("incorrect header check"); we route through a
# gzfile connection instead.
gzip_text <- function(text) {
  path <- tempfile()
  on.exit(unlink(path), add = TRUE)
  con <- gzfile(path, "wb")
  writeBin(charToRaw(text), con)
  close(con)
  readBin(path, what = "raw", n = file.info(path)$size)
}

# AWS Lambda sync responses are capped at 6 MB. Compress every response so
# macro outputs of ~25-50 MB raw can still fit. Callers detect the
# {encoding, payload} wrapper and decompress.
payload <- jsonlite::base64_enc(gzip_text(toJSON(result, auto_unbox = TRUE)))
cat(toJSON(list(encoding = "gzip+base64", payload = payload), auto_unbox = TRUE))
